Source/Task/LocklessQueue.h

```diff
@@ -0,0 +1,740 @@
+    // node elements are contained in a block.  Additional blocks will be allocated
+    // as needed but for best performance use a block size that does not require
+    // additional allocation for your work load. Passing zero will use the 
+    // minimum size.
+    //
+    LocklessQueue(_In_ uint32_t blockSize = 0x400) noexcept :
+        m_localHeap(*this),
+        m_activeList(*this),
+        m_heap(m_localHeap),
+        m_blockCache(nullptr)
+    {
+        m_localHeap.init(blockSize);
+        Initialize();
+    }
+
+    //
+    // Creates a new lockless queue.  The shared parameter provides another
+    // lockless queue that provides the heap for nodes. Queues that
+    // share the same node heap can share addresses. When a lockless queue
+    // that is sharing another queue's nodes is destroyed, any outstanding
+    // nodes in use are given back to the heap of the shared queue.  Destroying
+    // the host queue before destroying queues sharing its heap results in 
+    // undefined behavior.
+    //
+    LocklessQueue(_In_ LocklessQueue& shared) noexcept :
+        m_localHeap(*this),
+        m_activeList(*this),
+        m_heap(shared.m_heap),
+        m_blockCache(nullptr)
+    {
+        Initialize();
+    }
+
+    ~LocklessQueue() noexcept
+    {
+        if (&m_heap != &m_localHeap)
+        {
+            // This queue is sharing a heap, so
+            // we need to put all outstanding nodes
+            // back
+            TData data;
+            while(pop_front(data)) {}
+
+            // And the dummy node too.
+            Address dummy = m_activeList.head();
+            Node* node = to_node(dummy);
+            m_heap.free(node, dummy);
+        }
+    }
+
+    //
+    // Returns true if the queue is empty. Note that 
+    // since there are no locks this is a snapshot in time.
+    //
+    bool empty() noexcept
+    {
+        return m_activeList.empty();
+    }
+
+    //
+    // Reserves a link node in the queue, allocating it if
+    // necessary.  The node must be used in a later push_back
+    // or free_node call or it will be leaked.
+    //
+    bool reserve_node(_Out_ uint64_t& address) noexcept
+    {
+        Address a;
+        Node* n = m_heap.alloc(a);
+
+        if (n != nullptr)
+        {
+            address = a.value;
+            return true;
+        }
+        else
+        {
+            address = 0;
+            return false;
+        }
+    }
+
+    //
+    // Frees a previously reserved node.  Freeing a node
+    // that has been pushed results in undefined behavior. It
+    // will be hard to debug.
+    //
+    void free_node(_In_ uint64_t address) noexcept
+    {
+        Address a;
+        a.value = address;
+        Node* n = to_node(a);
+        m_heap.free(n, a);
+    }
+    
+    //
+    // Pushes the given data onto the back
+    // of the queue.  A copy is made of TData.
+    // If the push fails this returns false.
+    //
+    bool push_back(_In_ TData data) noexcept
```

Comment:
why not take TData by && to potentially avoid the copy? (it's now the caller responsibility to decide)

Reply:
Provided an overload below.